### PR TITLE
Fix `new` on compat-function constructors (#501, #502)

### LINKED ISSUE #501: Playground report: `function Foo(){}; new Foo();`
[Open in playground](https://www.gocciascript.dev/playground?share=eyJjb2RlIjoiZnVuY3Rpb24gRm9vKCl7XG4gIGNvbnNvbGUubG9nKCdoZWxsbycpO1xufTsgbmV3IEZvbygpOyIsIm1vZGUiOiJpbnRlcnByZXRlZCIsInJ1bm5lciI6ImV4ZWN1dGUiLCJhc2kiOnRydWUsImNvbXBhdFZhciI6ZmFsc2UsImNvbXBhdEZ1bmN0aW9uIjp0cnVlLCJ2ZXJzaW9uIjoibmlnaHRseSJ9)

| Setting | Value |
|---|---|
| Backend | `interpreted` |
| Runner | `execute` |
| Version | `nightly` |
| ASI | on |
| Compat `var` | off |
| Compat `function` | on |

---

### What did you expect to happen?

`function Foo(){}; new Foo();` should be create a constructor.

### What actually happened?

In interpreted mode only we get `TypeError: 'Foo' is not a constructor`, bytecode works exactly as expected with the constructor successfully created

### LINKED ISSUE #502: `new` on compat-function constructors: define expected behaviour and current gaps in both interpreter and bytecode
Follow-up to #501. That issue captures the interpreter "Foo is not a constructor" failure and notes that "bytecode works exactly as expected"; in practice bytecode is also broken, just in a different way. Spelling out the full picture across both modes so the expected behaviour is documented in one place and the engine can be fixed against a clear target.

### Interpreter (default for `GocciaScriptLoader` and `GocciaScriptLoaderBare`)

Every `new` on a plain function fails at the call site:

```js
function Foo() {}
new Foo();   // TypeError: 'Foo' is not a constructor
```

### Bytecode (`--mode=bytecode`)

The `new` call is accepted, but the result is `undefined` and `this` is unbound inside the body:

```js
function Foo() {}
console.log(typeof new Foo());                            // "undefined" — should be "object"

function Bar() {}
Bar.prototype.greet = function () { return "hi"; };
new Bar().greet();                                        // TypeError: Cannot read properties of undefined (reading 'greet')

function Baz(m) { this.m = m; }
const b = new Baz("x");                                   // TypeError: Cannot set properties of undefined (setting 'm')
```

So `new Foo()` returns `undefined` instead of a fresh ordinary object, and `this` inside the constructor body is `undefined` rather than the new instance.

## Expected behaviour (ES2026 §10.2.2 [[Construct]] / §10.2.1.1 OrdinaryCallBindThis)

When `new Foo(args)` is invoked on a function-keyword function under `--compat-function`:

1. A fresh ordinary object is created whose `[[Prototype]]` is `Foo.prototype` (PR #475 already wires the `prototype` own property).
2. The constructor body runs with `this` bound to that new ordinary object.
3. If the body explicitly returns an Object, that becomes the call result; otherwise the new ordinary object is the result.
4. `typeof new Foo()` is `"object"`, `new Foo() instanceof Foo` is `true`, methods set on `Foo.prototype` are visible on instances via the prototype chain.

Same shape that class declarations already implement, gated on `--compat-function` for declarations/expressions written with the `function` keyword.

## Why this matters

- **test262 compat** — `harness/sta.js` defines `function Test262Error(message) { this.message = ...; }` and tests do `new Test262Error(...)`. Without working `new` + `this`, downstream consumers (e.g. a test262.fyi engine entry) need a class shim in `runtime.js` that shadows the harness declaration.
- **Ergonomic compat** — the point of `--compat-function` is to let pre-class JS code run unchanged.
- **Subset-roadmap consistency** — `Foo.prototype` already exists (#475); `new` is the missing half of the constructor story.

## Acceptance

- [ ] Interpreter: `new Foo()` returns a fresh ordinary object with `Foo.prototype` as `[[Prototype]]`.
- [ ] Bytecode: `this` bound to the new instance inside the body; the call result is the new instance (or the body's explicit Object return value).
- [ ] Both modes: `instanceof Foo` is `true`, `typeof new Foo()` is `"object"`, prototype-attached methods are reachable on instances.
- [ ] Test added under `tests/language/function-keyword/` covering the four cases above in both modes; the "new on plain functions is not yet wired" caveat at `tests/language/function-keyword/prototype-property.js:88` can be removed.

### PR DESCRIPTION
**Pull request by frostney**
Fixes #501 and #502.

## Summary
- Both interpreter and bytecode VM now implement ES2026 §10.2.2 [[Construct]] for `function`-keyword constructors under `--compat-function`.
- `EvaluateNewExpression` (interpreter) and `ConstructValue` (bytecode VM) gain a `TGocciaFunctionBase` branch that allocates a fresh ordinary object whose `[[Prototype]]` is `Foo.prototype` (or `%Object.prototype%` per §10.1.14 when `prototype` is reassigned to a non-object), calls the body with the new object bound as `this`, and applies the spec return rules — explicit Object return wins, primitives are ignored.
- The bytecode side previously called the function with `undefined` as `this`; the interpreter side previously threw `TypeError: 'Foo' is not a constructor`.

## Tests
- New `tests/language/function-keyword/new-on-function.js` (23 tests): the four acceptance cases from #502 plus return-rule edge cases (object/primitive/null/undefined/array), spread args, nested `new`, prototype reassignment fallback, and TypeErrors for arrow / async / generator / concise method.
- Removed the "new on plain functions is not yet wired" caveat from `tests/language/function-keyword/prototype-property.js:168` per #502's acceptance list.

Fixes #501 and fixes #502 

## Test plan
- [x] `./build/GocciaTestRunner tests --asi --unsafe-ffi` — 8507/8507 pass
- [x] `./build/GocciaTestRunner tests --asi --unsafe-ffi --mode=bytecode` — 8507/8507 pass
- [x] `./format.pas --check`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: source/units/Goccia.Evaluator.pas [hunk 1]
```diff
@@ -3043,6 +3049,37 @@ function EvaluateNewExpression(const ANewExpression: TGocciaNewExpression; const
...
+        TGarbageCollector.Instance.AddTempRoot(ReceiverInstance);
+        try
+          ConstructResult := FunctionCallee.Call(Arguments, ReceiverInstance);
+          if ConstructResult is TGocciaObjectValue then
+            Result := ConstructResult
+          else
+            Result := ReceiverInstance;
+        finally
+          TGarbageCollector.Instance.RemoveTempRoot(ReceiverInstance);
+        end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.pas:3092]: Verified — reproduced the failure (`new Foo.bind(null, 1)(2)` threw `Cannot set properties of null` because `BoundFunction.Call(args, receiver)` discards the receiver and uses `FBoundThis` instead, and reading `prototype` from the bound wrapper returned `undefined`, falling back to `Object.prototype`).

Fixed in 8090e2da:
- Walk through `TGocciaBoundFunctionValue` wrappers to read `prototype` from the underlying [[BoundTargetFunction]] (handles doubly-bound too).
- Replace `FunctionCallee.Call` with `InvokeConstructableWithReceiver`, which already merges bound args and dispatches to the target.
- Added regression tests covering `new Foo.bind(null, 1)(2)`, prototype identity, doubly-bound chains, bound `this` being ignored, and TypeError for `new BoundArrow()`.

8512/8512 pass in interpreter and bytecode mode.

File: source/units/Goccia.VM.pas [hunk 2]
```diff
@@ -4542,14 +4544,39 @@ function TGocciaVM.ConstructValue(const AConstructor: TGocciaValue;
...
       if Assigned(TGocciaCallStack.Instance) then
-        TGocciaCallStack.Instance.Pop;
+        TGocciaCallStack.Instance.Push(ConstructorName, '', 0, 0);
+      try
+        ConstructResult := TGocciaFunctionBase(AConstructor).Call(
+          AArguments, ReceiverInstance);
+        if ConstructResult is TGocciaObjectValue then
+          Result := ConstructResult
+        else
+          Result := ReceiverInstance;
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.VM.pas]: Verified — same root cause as the interpreter-side comment (bound wrapper's `Call` ignores the passed-in receiver and uses `FBoundThis`).

Fixed in 8090e2da: `ConstructValue` now walks through `TGocciaBoundFunctionValue` wrappers to read `PROP_PROTOTYPE` from the underlying target, then dispatches via `InvokeConstructableWithReceiver` (which already does the bound-arg merging + recursive unwrapping). No need to recurse back into `ConstructValue` itself since `InvokeConstructableWithReceiver` covers Proxy/Native/Class/FunctionBase paths uniformly.

Regression tests covering bound construction added in `tests/language/function-keyword/new-on-function.js`.

File: source/units/Goccia.VM.pas [hunk 3]
```diff
@@ -4542,14 +4545,47 @@ function TGocciaVM.ConstructValue(const AConstructor: TGocciaValue;
...
-    finally
       if Assigned(TGocciaCallStack.Instance) then
-        TGocciaCallStack.Instance.Pop;
+        TGocciaCallStack.Instance.Push(ConstructorName, '', 0, 0);
+      try
+        // InvokeConstructableWithReceiver merges bound args, dispatches to
+        // the underlying target, and applies the spec return rules
+        // (explicit Object return wins; otherwise the receiver).
+        Result := InvokeConstructableWithReceiver(AConstructor, AArguments,
+          ReceiverInstance);
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.VM.pas:4600]: Verified — fixed in cc2d87e5. The reproducer (`new (Base.bind(null))()` where Base.constructor explicitly returns an object) confirmed the divergence in bytecode mode (`inst.swapped` was `undefined`).

Fix is broader than the literal suggestion because the VM has two class types:
- `TGocciaClassValue` — got the suggested `CallWithThisValue` + `ValidateClassConstructorReturn` + receiver-swap.
- `TGocciaVMClassValue` (bytecode-compiled classes) — has its constructor in `FConstructorValue`, not `FConstructorMethod`, so the suggested branch alone wouldn't fire. Added a `TGocciaVMClassValue` case ahead of it that routes through `Instantiate`, which already applies `ApplyOwnConstructorResult`'s receiver-swap rules.

Also exported `ValidateClassConstructorReturn` from the `Goccia.Evaluator` interface so the VM can call it.

Regression tests in `tests/language/function-keyword/new-on-function.js` cover bound class with explicit-object return and bound class with field initialization. 8514/8514 pass in both modes.

File: source/units/Goccia.Evaluator.pas [hunk 4]
```diff
@@ -3014,6 +3020,8 @@ function EvaluateNewExpression(const ANewExpression: TGocciaNewExpression; const
       CalleeName := TGocciaClassValue(Callee).Name
     else if Callee is TGocciaNativeFunctionValue then
       CalleeName := TGocciaNativeFunctionValue(Callee).Name
+    else if Callee is TGocciaFunctionBase then
+      CalleeName := TGocciaFunctionBase(Callee).GetProperty(PROP_NAME).ToStringLiteral.Value
```
[the remark this answers is not in the document]
Comment (frostney) [reply source/units/Goccia.Evaluator.pas:3024]: Verified — no fix needed.

`TGocciaFunctionBase` overrides `GetPropertyWithContext` (`Goccia.Values.FunctionBase.pas:155-169`) to intercept `PROP_NAME` before the generic property lookup, and unconditionally synthesizes a fresh `TGocciaStringLiteralValue.Create(GetFunctionName)`. Even an unnamed function returns `TGocciaStringLiteralValue.Create('')` — never nil. The nil-as-missing-property sentinel only applies to the generic `TGocciaObjectValue` fallback path that this branch never reaches because the `is TGocciaFunctionBase` guard already routed past it.

It's also the established idiom — the same unguarded `GetProperty(PROP_NAME).ToStringLiteral.Value` pattern is used in 6 other call sites against function values, including the adjacent `Goccia.VM.pas:4555` and `Goccia.Builtins.TestingLibrary.pas:1492,1535`. Adding a guard here would be inconsistent and would suggest a possibility that the type's contract precludes.

A user-deletable `name` property would matter for arbitrary objects, but `TGocciaFunctionBase` doesn't expose one — `name` is computed each call from the synthesized name slot.
